### PR TITLE
Add validator to avoid different subnets in the queues

### PR DESCRIPTION
**Pull request by chenwany**
Signed-off-by: chenwany <chenwany@amazon.com>

**Please See** [Git Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions)

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 1]
```diff
@@ -1098,6 +1098,19 @@ def right_scheduler_schema(self, data, **kwargs):
...
                 )
 
+    @validates_schema
+    def same_subnet_in_different_queues(self, data, **kwargs):
+        """Validate subnet_ids configured in different queues are the same."""
+        for queues in ["slurm_queues", "aws_batch_queues"]:
+            if data.get(queues, None):
+                queues_list = data.get(queues)
+                subnet_ids_list = []
+                for queue in queues_list:
```
Comment (charlesg3) [inline cli/src/pcluster/schemas/cluster_schema.py:1108]: could use `for queue in data[queues]`: here and avoid the extra variable created in line 1106

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 2]
```diff
@@ -1098,6 +1098,19 @@ def right_scheduler_schema(self, data, **kwargs):
                     f"Scheduling > *Queues section is not appropriate to the Scheduler: {configured_scheduler}."
                 )
 
+    @validates_schema
+    def same_subnet_in_different_queues(self, data, **kwargs):
+        """Validate subnet_ids configured in different queues are the same."""
+        for queues in ["slurm_queues", "aws_batch_queues"]:
+            if data.get(queues, None):
```
Comment (charlesg3) [inline cli/src/pcluster/schemas/cluster_schema.py:1105]: Feels more pythonic to use `if queue in data:` here

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 3]
```diff
@@ -1098,6 +1098,19 @@ def right_scheduler_schema(self, data, **kwargs):
...
+                queues_list = data.get(queues)
+                subnet_ids_list = []
```
Comment (charlesg3) [inline cli/src/pcluster/schemas/cluster_schema.py:1107]: if you make `subnet_ids` a `set` you can check `if len(subnet_ids) > 1` in the condition below (`if not all ...`) which would be shorter / read cleaner.

Comment (charlesg3) [reply cli/src/pcluster/schemas/cluster_schema.py:1107]: Would also be possible to do this all with a set-comprehension...

```.py
    def queue_has_subnet_ids(queue)
        return queue.networking and queue.networking.subnet_ids
    subnet_ids = {q.networking.subnet_ids for q in data.get(queues) if queue_has_subnet_ids(q)}
    if len(subnet_ids) > 1:
        raise ValidationError("SubnetIds configured in different queues should be the same.")
```

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 1 above)
Comment (chenwany) [reply cli/src/pcluster/schemas/cluster_schema.py:1108]: Good idea! Changed

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 2 above)
Comment (chenwany) [reply cli/src/pcluster/schemas/cluster_schema.py:1105]: I was using `data.get(queues, None)` to be aligned with the other validation function in our code, now change other places to use `if queues in data` to looks more clear

**Review by charlesg3 — APPROVED** [review]

**Pull request merged**